javav2/example_code/transcribe-streaming/src/main/java/com/amazonaws/transcribestreaming/BidirectionalStreaming.java

```diff
@@ -1,78 +1,78 @@
+import javax.sound.sampled.AudioInputStream;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.transcribestreaming.TranscribeStreamingAsyncClient;
+import software.amazon.awssdk.services.transcribestreaming.model.TranscribeStreamingException;
+import software.amazon.awssdk.services.transcribestreaming.model.StartStreamTranscriptionRequest;
+import software.amazon.awssdk.services.transcribestreaming.model.MediaEncoding;
+import software.amazon.awssdk.services.transcribestreaming.model.LanguageCode;
+import software.amazon.awssdk.services.transcribestreaming.model.StartStreamTranscriptionResponseHandler;
+import software.amazon.awssdk.services.transcribestreaming.model.TranscriptEvent;
+// snippet-end:[transcribe.java2.bidir_streaming.import]
+
+public class BidirectionalStreaming {
+
+    private static Object TranscribeStreamingException;
+
+    public static void main(String[] args) throws Exception {
+
+        Region region = Region.US_WEST_2;
+        TranscribeStreamingAsyncClient client = TranscribeStreamingAsyncClient.builder()
+                .region(region)
+                .build();
+
+        convertAudio(client);
+    }
+
+    // snippet-start:[transcribe.java2.bidir_streaming.main]
+    public static void convertAudio(TranscribeStreamingAsyncClient client) throws Exception {
+
+        try {
+
+            StartStreamTranscriptionRequest request = StartStreamTranscriptionRequest.builder()
+                    .mediaEncoding(MediaEncoding.PCM)
+                    .languageCode(LanguageCode.EN_US)
+                    .mediaSampleRateHertz(16_000).build();
+
+            TargetDataLine mic = com.amazonaws.transcribestreaming.Microphone.get();
+            mic.start();
+
+            AudioStreamPublisher publisher = new AudioStreamPublisher(new AudioInputStream(mic));
+
+            StartStreamTranscriptionResponseHandler response = StartStreamTranscriptionResponseHandler.builder()
+                    .subscriber(e -> {
+                        TranscriptEvent event = (TranscriptEvent) e;
+                        event.transcript().results()
+                                .forEach(r -> r.alternatives().forEach(a -> System.out.println(a.transcript())));
+                    }).build();
+
+            // Keeps Streaming until you end the Java program
+            client.startStreamTranscription(request, publisher, response);
+
+        } catch (TranscribeStreamingException e) {
+            System.err.println(e.awsErrorDetails().errorMessage());
+            System.exit(1);
+        }
+    }
+    // snippet-end:[transcribe.java2.bidir_streaming.main]
+
+    public static TargetDataLine get() throws Exception {
+        AudioFormat format = new AudioFormat(16000, 16, 1, true, false);
+        DataLine.Info datalineInfo = new DataLine.Info(TargetDataLine.class, format);
+
+        TargetDataLine dataLine = (TargetDataLine) AudioSystem.getLine(datalineInfo);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dataLine*. The resource is referenced at lines 73, 74. The resource is returned at line 75. There are other execution paths that don&#39;t close the resource or return it, for example, when TargetDataLine.open throws an exception. To prevent this resource leak, close *dataLine* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).